src/mlmc.py

```diff
@@ -0,0 +1,181 @@
+
+        self.count_new_num_of_simulations(new_num_of_simulations)
+
+    def count_new_num_of_simulations(self, num_of_simulations):
+        """
+        For each level counts further number of simulations by appropriate N
+        :param num_of_simulations: array      new number of simulations for each level
+        """
+
+        for step, level in enumerate(self._levels):
+            if self.number_of_simulations is not None:
+                num_of_simulations = self.number_of_simulations
+
+            if level.number_of_simulations < num_of_simulations[step]:
+                level.number_of_simulations = num_of_simulations[step] - level.number_of_simulations
+                # Launch further simulations
+                level.level()
+                level.number_of_simulations = num_of_simulations[step]
+
+    def create_level(self):
+        """
+        Create new level add its to the array of levels
+        Call method for counting number of simulation steps
+        Pass instance of Simulation to Level
+        """
+
+        level = Level(self.count_small_n(), self.simulation, self.moments_object)
+        self._levels.append(level)
+
+    def count_small_n(self):
+        """
+        Count number of steps for level
+        :return: array  [n_coarse, n_fine]
+        """
+        fine_steps = np.power(np.power((self.n_fine/self.n_coarse), (1/(self.number_of_levels-1))), self.index)*self.n_coarse
+
+        # Coarse number of simulation steps from previouse level
+        if len(self._levels) > 0:
+            level = self._levels[self.index - 1]
+            coarse_steps = level.n_ops_estimate()
+
+        else:
+            coarse_steps = 0
+        self.index += 1
+        return [np.round(coarse_steps).astype(int), np.round(fine_steps).astype(int)]
+
+    def count_num_of_sim_time(self):
+        """
+        For each level counts new N according to target_time
+        :return: array
+        """
+        num_of_simulations_time = []
+        amount = self.count_sum()
+
+        # Loop through levels
+        # Count new number of simulations for each level
+        for level in self._levels:
+            new_num_of_sim = np.round((self.target_time * np.sqrt(level.variance / level.n_ops_estimate()))
+                                      / amount).astype(int)
+
+            num_of_simulations_time.append(new_num_of_sim)
+        return num_of_simulations_time
+
+    def count_num_of_sim_variance(self):
+        """
+        For each level counts new N according to target_variance
+        :return: array
+        """
+        num_of_simulations_var = []
+        amount = self.count_sum()
+
+        # Loop through levels
+        # Count new number of simulations for each level
+        for level in self._levels:
+            new_num_of_sim_pom = []
+
+            for moment in level.moments:
+                new_num_of_sim_pom.append(np.round((amount * np.sqrt(np.abs(moment) / level.n_ops_estimate()))
```

Comment:
Aby tyto odhady fungovaly, potrebujeme pro jednotlive urovne odhad poctu operaci 'n_ops_estimate', to je ovsem zavisle jen na sim_steps (male n) tak na konkretni simulaci. Takze metoda 'n_ops_estimate' musi volat metodu simulace na fine urovni, ktera odhadne pocet operaci simulace.